### PR TITLE
MTC Pick and Place example

### PR DESCRIPTION
**Pull request by sjahr**
Adds example behavior and objective for Pick and Place example

### CODE REVIEW

File: src/tree_nodes_model.xml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            <MetadataFields>
+                <Metadata subcategory="Task Planning"/>
+            </MetadataFields>
+            <description>
+                <p>
+                    Wave hello with the end-effector
+                </p>
+            </description>
+            <inout_port name="task" default="{mtc_task}">MoveIt Task Constructor task.</inout_port>
+        </Action>
```
Comment (stephanie-eng) [inline src/tree_nodes_model.xml:31]: Is this relevant to this Behavior?

File: src/setup_mtc_place_from_pose.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                           false);
+    container->add(std::move(stage));
+  }
+
+  /** Detach Object **/
+  {
+    auto stage = std::make_unique<moveit::task_constructor::stages::ModifyPlanningScene>("Detach");
+    stage->detachObject(kSceneObjectName, kHandFrameName);
+    container->add(std::move(stage));
+  }
```
Comment (stephanie-eng) [inline src/setup_mtc_place_from_pose.cpp:206]: In the Pick From Pose behavior, we skipped attaching the object. Is this detach stage still necessary?

File: src/setup_mtc_place_from_pose.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+constexpr auto kEndEffectorName = "moveit_ee";
+constexpr auto kHandFrameName = "grasp_link";
+constexpr auto kHandOpenName = "open";
+constexpr auto kHandCloseName = "close";
+constexpr auto kApproachDistance = 0.1;
+constexpr auto kPropertyNameTrajectoryExecutionInfo = "trajectory_execution_info";
+constexpr double kIKTimeoutSeconds = 1.0;
+constexpr int kMaxIKSolutions = 20;
+constexpr auto kSceneObjectNameOctomap = "<octomap>";
+constexpr auto kSceneObjectName = "object";
```
Comment (stephanie-eng) [inline src/setup_mtc_place_from_pose.cpp:37]:
```suggestion
```
We got rid of this in pick, so probably can get rid of it here (if we remove the detach stage)

File: src/setup_mtc_place_from_pose.cpp [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    wrapper->setIgnoreCollisions(true);
+    container->add(std::move(wrapper));
+  }
+
+  /** Allow Collision
+      This stage allows collisions between the gripper and object for stages after this one (during the Close Hand,
+      Lift, and Retreat stages). */
+  {
+    auto stage =
+        std::make_unique<moveit::task_constructor::stages::ModifyPlanningScene>("Forbid collision (hand,object)");
```
Comment (stephanie-eng) [inline src/setup_mtc_place_from_pose.cpp:168]:
```suggestion
        std::make_unique<moveit::task_constructor::stages::ModifyPlanningScene>("Allow collision (hand,object)");
```

File: src/setup_mtc_pick_from_pose.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+constexpr auto kWorldFrame = "/world";
+constexpr auto kArmGroupName = "manipulator";
+constexpr auto kEndEffectorGroupName = "gripper";
+constexpr auto kEndEffectorName = "moveit_ee";
+constexpr auto kHandFrameName = "grasp_link";
+constexpr auto kHandCloseName = "close";
+constexpr auto kApproachDistance = 0.1;
+constexpr auto kPropertyNameTrajectoryExecutionInfo = "trajectory_execution_info";
+constexpr double kIKTimeoutSeconds = 1.0;
+constexpr int kMaxIKSolutions = 20;
```
Comment (sea-bass) [inline src/setup_mtc_pick_from_pose.cpp:34]: these 2 not being auto seems inconsistent, does it work with auto?

File: src/setup_mtc_pick_from_pose.cpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    stage->allowCollisions(kSceneObjectNameOctomap,
+                           task.value()
+                               ->getRobotModel()
+                               ->getJointModelGroup(kEndEffectorGroupName)
+                               ->getLinkModelNamesWithCollisionGeometry(),
+                           false);
+    container->insert(std::move(stage));
+  }
+
+  /** Generate the Inverse Kinematic (IK) solutions to move to the pose specified in the "grasp_pose" input port.
```
Comment (sea-bass) [inline src/setup_mtc_pick_from_pose.cpp:155]:
```suggestion
  /** Generate the Inverse Kinematics (IK) solutions to move to the pose specified in the "grasp_pose" input port.

File: src/setup_mtc_place_from_pose.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+constexpr auto kArmGroupName = "manipulator";
+constexpr auto kEndEffectorGroupName = "gripper";
+constexpr auto kEndEffectorName = "moveit_ee";
+constexpr auto kHandFrameName = "grasp_link";
+constexpr auto kHandOpenName = "open";
+constexpr auto kHandCloseName = "close";
+constexpr auto kApproachDistance = 0.1;
+constexpr auto kPropertyNameTrajectoryExecutionInfo = "trajectory_execution_info";
+constexpr double kIKTimeoutSeconds = 1.0;
+constexpr int kMaxIKSolutions = 20;
```
Comment (sea-bass) [inline src/setup_mtc_place_from_pose.cpp:35]: same here re: using `auto`

File: src/setup_mtc_place_from_pose.cpp [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+{
+const auto kLogger = rclcpp::get_logger("SetupMTCPickFromPose");
+using MoveItErrorCodes = moveit_msgs::msg::MoveItErrorCodes;
+
+// Port names for input and output ports.
+constexpr auto kPortIDTask = "task";
+constexpr auto kPortIDGraspPose = "place_pose";
+
+// Behavior constants
+constexpr auto kWorldFrame = "/world";
```
Comment (sea-bass) [inline src/setup_mtc_place_from_pose.cpp:25]: this shouldn't have a leading `/`

File: src/setup_mtc_pick_from_pose.cpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+{
+const auto kLogger = rclcpp::get_logger("SetupMTCPickFromPose");
+using MoveItErrorCodes = moveit_msgs::msg::MoveItErrorCodes;
+
+// Port names for input and output ports.
+constexpr auto kPortIDTask = "task";
+constexpr auto kPortIDGraspPose = "grasp_pose";
+
+// Behavior constants
+constexpr auto kWorldFrame = "/world";
```
Comment (sea-bass) [inline src/setup_mtc_pick_from_pose.cpp:25]: this shouldn't have a leading `/`

File: src/setup_mtc_place_from_pose.cpp [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    geometry_msgs::msg::Vector3Stamped approach_vector_msg;
+    tf2::toMsg(approach_vector, approach_vector_msg.vector);
+    approach_vector_msg.header.frame_id = kHandFrameName;
+
+    stage->setDirection(approach_vector_msg);
+    stage->setTimeout(10);
+    container->add(std::move(stage));
+  }
+
+  /** Generate the Inverse Kinematic (IK) solutions to move to the pose specified in the "place_pose" input port.
```
Comment (sea-bass) [inline src/setup_mtc_place_from_pose.cpp:138]:
```suggestion
  /** Generate the Inverse Kinematics (IK) solutions to move to the pose specified in the "place_pose" input port.

**Review by sea-bass — COMMENTED** [review]
only found small things

File: src/setup_mtc_place_from_pose.cpp [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+                              boost::any_cast<moveit::task_constructor::TrajectoryExecutionInfo>(
+                                  task.value()->properties().get(kPropertyNameTrajectoryExecutionInfo)));
+  container->properties().configureInitFrom(moveit::task_constructor::Stage::PARENT);
+
+  container->setProperty("group", kArmGroupName);
+  container->setProperty("hand", kEndEffectorGroupName);
+  container->setProperty("eef", kEndEffectorName);
+  container->setProperty("ik_frame", kHandFrameName);
+
+  /** Set Allowed Collisions since the hand holds an collision object */
```
Comment (MikeWrock) [inline src/setup_mtc_place_from_pose.cpp:92]:
```suggestion
  /** Allow Collision since the hand holds a collision object */
```

**Review by MikeWrock — APPROVED** [review]

File: src/setup_mtc_pick_from_pose.cpp (hunk 9 above)
Comment (sjahr) [inline src/setup_mtc_pick_from_pose.cpp:25]:
```suggestion
constexpr auto kWorldFrame = "world";
```

File: src/setup_mtc_pick_from_pose.cpp (hunk 5 above)
Comment (sjahr) [inline src/setup_mtc_pick_from_pose.cpp:34]:
```suggestion
constexpr auto kIKTimeoutSeconds = 1.0;
constexpr auto kMaxIKSolutions = 20;
```

File: src/setup_mtc_place_from_pose.cpp (hunk 8 above)
Comment (sjahr) [inline src/setup_mtc_place_from_pose.cpp:25]:
```suggestion
constexpr auto kWorldFrame = "world";
```

File: src/setup_mtc_place_from_pose.cpp (hunk 7 above)
Comment (sjahr) [inline src/setup_mtc_place_from_pose.cpp:35]:
```suggestion
constexpr auto kIKTimeoutSeconds = 1.0;
constexpr auto kMaxIKSolutions = 20;
```
